### PR TITLE
[WIP] [NN] Refactor NN package

### PR DESCRIPTION
**Pull request by jermainewang**
## Description

This PR refactor the out-of-date NN package. Some guidelines:
* We won't try to unify different backend for this. User needs to explicitly import, for example `dgl.nn.pytorch`. This is because the interface might be quite different for different backends.
* Try to follow the convention of each backend.
* Don't forget to add docstring.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR

## Changes

- [x] `dgl.nn.pytorch.conv.GraphConv`
- [x] `dgl.nn.pytorch.softmax.EdgeSoftmax`
- [x] `dgl.nn.mxnet.conv.GraphConv`
- [x] Also change the example to use these predefined modules.

## Additional Note
- I've also done some small fixes on the way to modify the examples
- The incorporation of `nn` modules into the examples seem to have slowed down the speed of execution as there are some redundant operations of setting and popping data.

### CODE REVIEW

**Comment by mufeili** [comment]
@jermainewang I assume only the examples will be affected and the tutorials are left unchanged?

**Comment by jermainewang** [comment]
Yes.

**Comment by mufeili** [comment]
@jermainewang I'm done and you can take a look.

**Comment by mufeili** [comment]
@jermainewang I'm done and you may take a look. There are some additional changes that I did not realize until now (the ones in `sampler.h`, `sampler.cc`, `pylintrc` and `task_lint.sh`), I suppose these are your changes?

File: python/dgl/nn/pytorch/softmax.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+        A field name such that g.ndata[max_logits_name] will store
+        :math:`\max_{j\in\mathcal{N}(i)}z_{ij}`.
+    normalizer_name : str
+        A field name such that g.ndata[normalizer_name] will store
+        :math:`\sum_{j\in\mathcal{N}(i)}exp(z_{ij})`
+    """
+    def __init__(self,
+                 logits_name,
+                 max_logits_name,
+                 normalizer_name):
```
Comment (jermainewang) [inline python/dgl/nn/pytorch/softmax.py]: Give default names for them?

File: python/dgl/nn/pytorch/softmax.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+            The input edge feature
+        graph : DGLGraph
+            The graph.
+
+        Returns
+        -------
+        Unnormalized scores : torch.Tensor
+            This part gives :math:`\exp(z_{ij})`'s
+        Normalizer : torch.Tensor
+            This part gives :math:`\sum_{j\in\mathcal{N}(i)}\exp(z_{ij})`
```
Comment (jermainewang) [inline python/dgl/nn/pytorch/softmax.py]: I guess eventually we want to just return the normalized scores as normal softmax does. But I'm fine with this right now. We could change this when we have our kernel to do that more easily.

File: python/dgl/nn/pytorch/softmax.py (hunk 1 above)
Comment (mufeili) [reply python/dgl/nn/pytorch/softmax.py]: Though the chance is not very high, I want to avoid any potential risk of undesired overwriting.

Comment (jermainewang) [reply python/dgl/nn/pytorch/softmax.py]: We could also detect whether the name has been taken and use another one. Then those temporary variables will not be seen by the user.

**Comment by jermainewang** [comment]
> There are some additional changes that I did not realize until now (the ones in `sampler.h`, `sampler.cc`, `pylintrc` and `task_lint.sh`), I suppose these are your changes?

Yes

File: python/dgl/nn/pytorch/softmax.py (hunk 1 above)
Comment (mufeili) [reply python/dgl/nn/pytorch/softmax.py]: Done.

File: python/dgl/nn/mxnet/conv.py [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+        The learnable bias tensor.
+    """
+    def __init__(self,
+                 in_feats,
+                 out_feats,
+                 norm=True,
+                 dropout=0.,
+                 bias=False,
+                 activation=mx.nd.relu,
+                 feat_name="_gconv_feat"):
```
Comment (jermainewang) [inline python/dgl/nn/mxnet/conv.py]: Let's not put dropout and activation here. My idea is to make it similar to torch.nn.conv2d. Activation and dropout can be handled outside.

File: python/dgl/nn/pytorch/softmax.py (hunk 1 above)
Comment (jermainewang) [reply python/dgl/nn/pytorch/softmax.py]: Then, it's totally safe to remove those "_name" args from the arglist. To think it as a normal pytorch module, the internal local variables are not exposed at all. So are these names. We could also change `GraphConv` in the same way. What do you think?

Comment (mufeili) [reply python/dgl/nn/pytorch/softmax.py]: Yeah, I agree. I just forgot to make those changes. Thanks.

File: python/dgl/nn/mxnet/conv.py (hunk 3 above)
Comment (mufeili) [reply python/dgl/nn/mxnet/conv.py]: My motivation is to make this more like Kipf's initial implementation of [GraphConvolutionLayer](https://github.com/tkipf/gcn/blob/master/gcn/layers.py#L132) where both dropout and activation are part of it. This also make potential model code more compact and expressive like `nn.Linear(bias=True)` is more compact than performing matrix multiplication, check whether bias is to be added and then perform a bias addition if needed.

Comment (jermainewang) [reply python/dgl/nn/mxnet/conv.py]: Ok, to me this is a more [Tensorflow/Keras way](https://keras.io/layers/core/), but I'm think it's not a big deal. One thing is that you might want to set `activation=None` by default, which is more commonly seen.

Comment (jermainewang) [reply python/dgl/nn/mxnet/conv.py]: Also, dropout is actually a tricky stuff. Where is the right place to apply it? On the input? On the aggregated value? Or after activation? I feel there is so many decisions so it's good to leave it to users. (TF/Keras also does not include dropout).

Comment (mufeili) [reply python/dgl/nn/mxnet/conv.py]: Done.

**Comment by jermainewang** [comment]
Did some minor fix on MXNet side. Thanks Murphy!
